### DISCUSSION TITLE
can binaryen be used to turn "old" JS style exceptions into new wasm exceptions?

### DISCUSSION THREAD

**Discussion by rikogeln-havok** [WebAssembly/binaryen#5409] [Q&A] [answered]
Hi everyone - 
In a unity webGL project I see lots of invoke_<signature>/dyncall_<signature> pairs for exception handling. 
Since it's a Unity project and unity ships with an old version of emscripten/llvm (i think it's 2.0.19) I wanted to check if it's possible to use an additional step of us running binaryen on the produced .wasm file and turn js style exceptions into the new wasm-exceptions. 

I know that the ideal way would be to get an updated version of emscripten/llvm and somehow convice unity to pass on the -fwasm-exception flag. Before I do this I wanted to see if we can do this in an "external" postbuild step. 

Thanks,
r

**Comment by tlively** [comment, marked as answer]
It might be possible in principle, but that hasn't been implemented.

**Reply by rikogeln-havok** [reply]
thanks for the reply @tlively !
Do you think it's reasonable for someone unfamiliar to the binaryen codebase to dive in and implement it in some weeks?

**Reply by tlively** [reply]
It's hard to say. If you start with a good understanding of how the code needs to be transformed, then figuring out how to do that transformation in a Binaryen pass probably wouldn't be so bad. But if you're hoping to figure out the correct transformation as you go, then I would expect it to be extremely difficult.

**Reply by kripken** [reply]
I believe this might be doable in some weeks. Each invoke call could be replaced with a try-catch around a normal call. There would just be a single tag type, which contains a single integer (the thrown pointer). The only other thing you'd need to change is the actual throwing (to create a tag and throw it), which calls out to JS to throw - but that would require a hack of some kind (finding the right import in a minified build is not trivial, as the imports aren't marked).

**Reply by rikogeln-havok** [reply]
Thanks for the replies people! 
I'll explore the other options further (using a different emscripten/llvm version in unity and passing -fwasm-exceptions in the correct places of the unity build system) and if they fail come back to this if necessary. 

Also so far I cannot prove that we'd get lots of perf improvements with wasm exceptions but estimate the upper bound to be around 2% or so 

thanks again and thanks for your work on binaryen!
